pkg/controller/keyvault/gateway_secret_provider_class.go

```diff
@@ -0,0 +1,244 @@
+package keyvault
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/Azure/aks-app-routing-operator/pkg/config"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/controllername"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/metrics"
+	"github.com/Azure/aks-app-routing-operator/pkg/manifests"
+	"github.com/Azure/aks-app-routing-operator/pkg/util"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+	secv1 "sigs.k8s.io/secrets-store-csi-driver/apis/v1"
+)
+
+var (
+	gatewaySecretProviderControllerName = controllername.New("keyvault", "gateway", "secret", "provider")
+)
+
+const (
+	saNameUriErr = "`App Routing requires a Certificate URI as well as a ServiceAccount that references an Identity that has access to your Keyvault in order to pull TLS certificates. Make sure both are specified in your Listener's TLS options for all listeners that require TLS certs."
+)
+
+// GatewaySecretProviderClassReconciler manages a SecretProviderClass for Gateway resource that specifies a ServiceAccount
+// and Keyvault URI in its TLS options field. The SPC is used to mirror the Keyvault values into
+// a k8s secret so that it can be used by the CRD controller.
+type GatewaySecretProviderClassReconciler struct {
+	client client.Client
+	events record.EventRecorder
+	config *config.Config
+}
+
+func NewGatewaySecretClassProviderReconciler(manager ctrl.Manager, conf *config.Config) error {
+	if conf.DisableKeyvault {
+		return nil
+	}
+	metrics.InitControllerMetrics(gatewaySecretProviderControllerName)
+
+	return gatewaySecretProviderControllerName.AddToController(
+		ctrl.
+			NewControllerManagedBy(manager).
+			For(&gatewayv1.Gateway{}), manager.GetLogger(),
+	).Complete(&GatewaySecretProviderClassReconciler{
+		client: manager.GetClient(),
+		events: manager.GetEventRecorderFor("aks-app-routing-operator"),
+		config: conf,
+	})
+}
+
+func (g *GatewaySecretProviderClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	res := ctrl.Result{}
```

Comment:
change this pattern for all new controllers

Reply:
yeah will do! expanding pod placeholder tests then pushing a commit addressing all the above